cmd/packer-plugin-check/main.go

```diff
@@ -0,0 +1,177 @@
+	// TODO: this should be updated once we have defined what's going to be for plguin's docs
+	wd, err := os.Getwd()
+	if err != nil {
+		return err
+	}
+
+	docsDir := wd + "/docs"
+	stat, err := os.Stat(docsDir)
+	if err != nil {
+		return fmt.Errorf("could not find docs folter: %s", err.Error())
+	}
+	if !stat.IsDir() {
+		return fmt.Errorf("expecting docs do be a directory of mdx files")
+	}
+
+	var mdxFound bool
+	_ = filepath.Walk(docsDir, func(path string, info os.FileInfo, err error) error {
+		if !info.IsDir() && filepath.Ext(path) == ".mdx" {
+			mdxFound = true
+			return io.EOF
+		}
+		return nil
+	})
+
+	if mdxFound {
+		fmt.Printf("a mdx file was found inside the docs folder\n")
+		return nil
+	}
+	return fmt.Errorf("no docs files found, make sure to have the docs in place before releasing")
+}
+
+// checkPluginName checks for the possible valid names for a plugin, packer-plugin-* or packer-[builder|provisioner|post-processor]-*.
+// If the name is prefixed with `packer-[builder|provisioner|post-processor]-`, packer won't be able to install it,
+// therefore a WARNING will be shown.
+func checkPluginName(name string) error {
+	if strings.HasPrefix(name, "packer-plugin-") {
+		return nil
+	}
+	if strings.HasPrefix(name, "packer-builder-") ||
+		strings.HasPrefix(name, "packer-provisioner-") ||
+		strings.HasPrefix(name, "packer-post-processor-") {
+		fmt.Printf("\n[WARNING] Plugin is named with old prefix `packer-[builder|provisioner|post-processor]-{name})`. " +
+			"These will be detected but Packer cannot install them automatically. " +
+			"Update the name to packer-plugin-{name} and the plugin will become available through `packer init` command.\n")
+		return nil
+	}
+	return fmt.Errorf("plugin's name is not valid")
+}
+
+// discoverAndLoad will discover the plugin binary from the current directory and load any builder/provisioner/post-processor
+// in the plugin configuration. At least one builder, provisioner, or post-processor should be found to template the plugin's
+// compatibility with Packer.
+func discoverAndLoad() error {
+	config := plugin.Config{
+		PluginMinPort: 10000,
+		PluginMaxPort: 25000,
+	}
+	err := config.Discover()
+	if err != nil {
+		return err
+	}
+
+	// TODO: template correctness of plugins loaded by checking them against the output of the `describe` command.
+	builders, provisioners, postProcessors := config.GetPlugins()
+	if len(builders) == 0 &&
+		len(provisioners) == 0 &&
+		len(postProcessors) == 0 {
+		return fmt.Errorf("couldn't load any Builder/Provisioner/Post-Processor from the plugin binary")
+	}
+
+	return checkHCL2ConfigSpec(builders, provisioners, postProcessors)
+}
+
+// checkHCL2ConfigSpec checks if the hcl2spec config is present for the given plugins by validating that ConfigSpec() does not
+// return an empty map of specs.
+func checkHCL2ConfigSpec(builders packersdk.MapOfBuilder, provisioners packersdk.MapOfProvisioner, postProcessors packersdk.MapOfPostProcessor) error {
```

Comment:
@nywilken Now we will validate that there is a configuration for each of the plugins.
It won't validate if the configuration is correct though. But it is good that if in the future we don't use more hcl2spec.go files this will still work 🤔

Reply:
ho really cool :D Though I think that the compile time error should be sufficient here 🤔 .

Reply:
Though I think that starting the plugin and asking for the spec is a very nice check !

Reply:
The situation I created while testing this, was the one where I forgot to generate the hcl2spec file and therefore ConfigSpec was returning nil. In this case, the plugin is loaded but it won't be able to read hcl2 and this is possible to happen because it won't fail at compilation time if one is testing with JSON for example.

Reply:
Ah yes, that totally makes sense, make the compiler stop complaining and then forget about it ! ( I put panics in my TODO funcs to remind myself of doing them so I kinda forgot about that case 🤦🏼 )